discovery/discovery.go

```diff
@@ -63,7 +63,7 @@ func ProvidersFromConfig(cfg config.ServiceDiscoveryConfig, logger log.Logger) m
 		app("dns", i, dns.NewDiscovery(c, log.With(logger, "discovery", "dns")))
 	}
 	for i, c := range cfg.FileSDConfigs {
-		app("file", i, file.NewDiscovery(c, log.With(logger, "discovery", "file")))
+		app("file", i, file.NewDiscovery(c, log.With(logger, "discovery", "file"), i))
```

Comment:
This is only the index within one scrape_config, which may change across reloads if the user reorders things.

What I'd suggest doing is having each file discovery register with TimestampCollector. Each file discovery would also remember the mtimes of its files. Then on Collect the TimestampCollector would pull in all this information, and de-dupe.